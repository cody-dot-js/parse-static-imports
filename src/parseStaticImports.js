import parseModuleName from "./parseModuleName";
import parseStarImport from "./parseStarImport";
import parseDefaultImport from "./parseDefaultImport";
import parseNamedImports from "./parseNamedImports";
import isSideEffectOnly from "./isSideEffectOnly";
import removeLineComments from "./removeLineComments";
import removeBlockComments from "./removeBlockComments";

function preprocess(file = "") {
  const normalizeLineEndings = file.replace(/\r/gm, "").trim();

  if (!file.includes("import ")) {
    return [];
  }

  return normalizeLineEndings
    .split("import ")
    .filter(Boolean)
    .map(str => {
      const normalized = str.replace(/(\'|\`)/gm, '"'); // eslint-disable-line no-useless-escape
      const openImportQuoteIdx = normalized.indexOf('"');
      const closeImportQuoteIdx = normalized.indexOf(
        '"',
        openImportQuoteIdx + 1
      );

      const restOfImport = normalized
        .substring(0, closeImportQuoteIdx + 1)
        .trim();

      return `import ${restOfImport}`;
    })
    .map(removeLineComments)
    .map(removeBlockComments);
}

/** parseStaticImports
 * returns a a list of objects whose properties represent significant elements
 * of the static import.
 *
 * @param {String} file - Contents of a file containing static imports
 * @returns {Object[]} List of static imports found in the given file contents
 */
export default function parseStaticImports(file = "") {
  return preprocess(file).reduce((acc, str) => {
    const sideEffectOnly = isSideEffectOnly(str);

    return [
      ...acc,
      {
        moduleName: parseModuleName(str),
        defaultImport: sideEffectOnly ? "" : parseDefaultImport(str),
        namedImports: sideEffectOnly ? [] : parseNamedImports(str),
        starImport: sideEffectOnly ? "" : parseStarImport(str),
        sideEffectOnly
      }
    ];
  }, []);
}
