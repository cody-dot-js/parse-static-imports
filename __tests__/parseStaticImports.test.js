import fs from "fs";
import path from "path";
import { promisify } from "util";
import parseStaticImports from "../src/parseStaticImports";
import importSyntaxList from "./__fixtures__/importSyntaxList";

const readFileAsync = promisify(fs.readFile);
const fixturesDirectoryPath = path.resolve(__dirname, "__fixtures__");

test("when given nothing returns an empty list", () => {
  const expected = [];
  const result = parseStaticImports();

  expect(result).toStrictEqual(expected);
});

describe("for each example in the import syntax list it returns the expected shape", () => {
  importSyntaxList.forEach(({ actual, expected }) => {
    const result = parseStaticImports(actual);

    it(actual, () => {
      expect(result).toStrictEqual(expected);
    });
  });
});

test("it matches the snapshot for each app example fixture", async () => {
  const appExamplesPath = path.resolve(fixturesDirectoryPath, "app-examples");

  const appExamplePaths = fs
    .readdirSync(appExamplesPath)
    .map(filename => path.join(appExamplesPath, filename));

  const appExamples = await Promise.all(
    appExamplePaths.map(fixturePath => readFileAsync(fixturePath, "utf8"))
  );

  appExamples.forEach((fixture, index) => {
    const result = {
      result: parseStaticImports(fixture),
      name: path.parse(appExamplePaths[index]).base
    };

    expect(result).toMatchSnapshot();
  });
});

test("when given gibberish it returns an empty list", () => {
  const expected = [];
  const gibberish =
    "Esse adipisicing ad magna sunt exercitation deserunt labore.";
  const result = parseStaticImports(gibberish);

  expect(result).toStrictEqual(expected);
});
