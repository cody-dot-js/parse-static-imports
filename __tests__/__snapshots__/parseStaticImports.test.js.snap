// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches the snapshot for each app example fixture 1`] = `
Object {
  "name": "App.jsx",
  "result": Array [
    Object {
      "defaultImport": "React",
      "moduleName": "react",
      "namedImports": Array [
        Object {
          "alias": "foo",
          "name": "useState",
        },
        Object {
          "alias": "useEffect",
          "name": "useEffect",
        },
        Object {
          "alias": "useLayoutEffect",
          "name": "useLayoutEffect",
        },
        Object {
          "alias": "useRef",
          "name": "useRef",
        },
        Object {
          "alias": "useReducer",
          "name": "useReducer",
        },
      ],
      "sideEffectOnly": false,
      "starImport": "",
    },
    Object {
      "defaultImport": "",
      "moduleName": "react-dom",
      "namedImports": Array [
        Object {
          "alias": "render",
          "name": "render",
        },
      ],
      "sideEffectOnly": false,
      "starImport": "",
    },
    Object {
      "defaultImport": "",
      "moduleName": "./logo.svg",
      "namedImports": Array [],
      "sideEffectOnly": false,
      "starImport": "logos",
    },
    Object {
      "defaultImport": "",
      "moduleName": "./App.css",
      "namedImports": Array [],
      "sideEffectOnly": true,
      "starImport": "",
    },
  ],
}
`;

exports[`it matches the snapshot for each app example fixture 2`] = `
Object {
  "name": "index.jsx",
  "result": Array [
    Object {
      "defaultImport": "React",
      "moduleName": "react",
      "namedImports": Array [],
      "sideEffectOnly": false,
      "starImport": "",
    },
    Object {
      "defaultImport": "",
      "moduleName": "react-dom",
      "namedImports": Array [
        Object {
          "alias": "render",
          "name": "render",
        },
      ],
      "sideEffectOnly": false,
      "starImport": "",
    },
    Object {
      "defaultImport": "",
      "moduleName": "./index.css",
      "namedImports": Array [],
      "sideEffectOnly": true,
      "starImport": "",
    },
    Object {
      "defaultImport": "App",
      "moduleName": "./App",
      "namedImports": Array [],
      "sideEffectOnly": false,
      "starImport": "",
    },
    Object {
      "defaultImport": "",
      "moduleName": "./serviceWorker",
      "namedImports": Array [],
      "sideEffectOnly": false,
      "starImport": "serviceWorker",
    },
    Object {
      "defaultImport": "",
      "moduleName": "./really/long/import/ea/ut/exercitation/commodo/ad/ea/reprehenderit/nisi/deserunt/sint/ipsum/consectetur/occaecat/dolor/minim/dolor/consectetur/ad/reprehenderit/aute/nisi/duis/labore/eu/reprehenderit/deserunt/incididunt/quis/deserunt/ex/incididunt/occaecat/magna/veniam/sunt/anim/id/qui/non/consectetur/est/ad/laboris/velit/non/irure/cillum/sint/consectetur/cillum/deserunt/id/do/dolor/labore/pariatur/qui/est/qui/incididunt/dolore/eu/cupidatat/ea/pariatur/aliqua/laboris/excepteur/eiusmod/culpa/voluptate/ullamco/sint/dolor/do/incididunt/aliquip/irure/id/in/occaecat/duis/id/et/magna/id/enim/officia/aliquip/exercitation/cupidatat/quis/ipsum/non/incididunt/adipisicing/voluptate/quis/nostrud/veniam/magna/esse/excepteur/officia/lorem/ipsum/officia/fugiat/proident/amet/culpa/esse/amet/sit/esse/dolor/commodo/voluptate/excepteur/labore/cupidatat/ad/eu/ad/dolor/id/aliqua/in/veniam/laboris/excepteur/exercitation/officia/aliquip/elit/aliquip/dolore/sit/fugiat/fugiat/officia/fugiat/commodo/amet/laboris/nostrud",
      "namedImports": Array [],
      "sideEffectOnly": true,
      "starImport": "",
    },
  ],
}
`;
